lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,471 @@
+            ) and assigned_cert["certificate"] != self._pull_tls_file_from_workload(File.CERT):
+                self._push_tls_file_to_workload(File.CERT, assigned_cert["certificate"])
+                self._push_tls_file_to_workload(File.CA, assigned_cert["ca"])
+                logger.info(
+                    "TLS: Certificate from access relation saved for unit %s.",
+                    self.charm.unit.name,
+                )
+                self._reload_vault_container()
+                return
+
+        elif not self._tls_file_pushed_to_workload(
+            File.CA
+        ) or not self._tls_file_pushed_to_workload(File.CERT):
+            self._generate_self_signed_certs(ingress_address)
+            logger.info(
+                "TLS: Saved Vault generated CA and self signed certificate to %s.",
+                self.charm.unit.name,
+            )
+            self._reload_vault_container()
+
+    def send_ca_cert(self):
+        """Sends the existing CA cert in the workload to all relations."""
+        if ca := self._pull_tls_file_from_workload(File.CA):
+            for relation in self.charm.model.relations.get(SEND_CA_CERT_RELATION_NAME, []):
+                self.certificate_transfer.set_certificate(
+                    certificate="", ca=ca, chain=[], relation_id=relation.id
+                )
+            logger.info("TLS: Sent CA certificate to other relations")
+        else:
+            for relation in self.charm.model.relations.get(SEND_CA_CERT_RELATION_NAME, []):
+                self.certificate_transfer.remove_certificate(relation.id)
+            logger.info("TLS: Removed CA cert from relations")
+
+    def _generate_self_signed_certs(self, ingress_address: str) -> None:
+        """Recreates a unit certificate from the Vault CA certificate, then saves it.
+
+        Args:
+            ingress_address: The subject of the unit certificate.
+        """
+        self._remove_all_certs_from_workload()
+
+        if not (private_key := self._pull_tls_file_from_workload(File.KEY)):
+            private_key = generate_private_key().decode()
+            self._push_tls_file_to_workload(File.KEY, private_key)
+
+        ca_private_key, ca_certificate = self._get_ca_certificate_secret_in_peer_relation()
+        self._push_tls_file_to_workload(File.CA, ca_certificate)
+        sans_ip = [ingress_address]
+        certificate = generate_vault_unit_certificate(
+            subject=ingress_address,
+            sans_ip=sans_ip,
+            sans_dns=[socket.getfqdn()],
+            ca_certificate=ca_certificate.encode(),
+            ca_private_key=ca_private_key.encode(),
+            unit_private_key=private_key.encode(),
+        )
+        self._push_tls_file_to_workload(File.CERT, certificate)
+
+    def _new_certificate_request_pending(self) -> bool:
+        """Check if we have a pending CSR by comparing the requirer and provider databags."""
+        # TODO: This will be a library function in the future
```

Comment:
Can this TODO be removed?

Reply:
Not yet, my PR for this feature is up right now in the interface repo, but I'd rather keep this here so I don't forget to actually put it in here after it's merged. I can remove it and write it somewhere else, but there are a bunch of TODO tags in other libraries so I thought it would be ok.

Reply:
> my PR for this feature is up right now in the interface repo

Can you link to this? I was trying to follow this discussion and I can't find it.

Reply:
No need, the changes are merged in the other repo and this function is now updated. The TODO is removed.